File: userland/capsule_text_editor/src/editor/ribbon/cells.rs
// NONOS Operating System
// Copyright (C) 2026 NONOS Contributors
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

//! The left-to-right cell walk. Pill widths follow their measured labels, so a
//! longer style name pushes the toggles along and the hit-test follows, because
//! it reads back the very cells this produced.

use alloc::vec::Vec;
use nonos_app_skeleton::measure_ttf;

use super::items::{RibbonItem, ICON_COUNT, TOGGLES};
use super::metrics::{RibbonCell, BAR_PAD, CHEVRON_W, GAP, PILL_PAD, SEP_GAP, SQUARE_W};
use crate::editor::layout::{ACTIVITY_W, CHROME_PX};

pub(super) fn cells(labels: &[&str; 3]) -> Vec<RibbonCell> {
    let mut out = Vec::new();
    let mut x = ACTIVITY_W + BAR_PAD;
    for (p, label) in labels.iter().enumerate() {
        let w = measure_ttf(label, CHROME_PX).max(0) as u32 + PILL_PAD * 2 + CHEVRON_W;
        out.push(RibbonCell { x0: x, x1: x + w, item: RibbonItem::Pill(p) });
        x += w + GAP;
    }
    x += SEP_GAP;
    for t in 0..TOGGLES.len() {
        out.push(RibbonCell { x0: x, x1: x + SQUARE_W, item: RibbonItem::Toggle(t) });
        x += SQUARE_W + GAP;
    }
    x += SEP_GAP;
    for k in 0..ICON_COUNT {
        out.push(RibbonCell { x0: x, x1: x + SQUARE_W, item: RibbonItem::Icon(k) });
        x += SQUARE_W + GAP;
    }
    out
}

File: scripts/bootmatrix/cells.py
# NONOS Operating System
# Copyright (C) 2026 NONOS Contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.
"""The machines the kernel has to boot on.

One cell is one QEMU configuration. The table is the claim: every shipping
image boots on each of these, every time, and the runner's job is to find the
one that does not. `profile` names which built ESP the cell boots, since the
multiprocessor image is a separate build of the same tree.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Cell:
    name: str
    profile: str
    machine: str
    cpus: int
    iommu: bool = False
    # Kill QEMU once the store is serving, then boot the same disk again and
    # require it to come back clean. Crash consistency of the block store.
    kill: bool = False

    def wants_smp_proof(self):
        return self.cpus > 1


CELLS = [
    Cell("q35-up", "up", "q35", 1),
    Cell("q35-smp2", "smp", "q35", 2),
    Cell("q35-smp4", "smp", "q35", 4),
    Cell("q35-smp8", "smp", "q35", 8),
    Cell("i440fx-up", "up", "pc", 1),
    Cell("i440fx-smp4", "smp", "pc", 4),
    Cell("q35-iommu-up", "up", "q35", 1, iommu=True),
    Cell("q35-iommu-smp4", "smp", "q35", 4, iommu=True),
    Cell("q35-kill-reboot", "up", "q35", 1, kill=True),
]


def select(names):
    """The cells named, in table order, or all of them when nothing is named."""
    if not names:
        return list(CELLS)
    known = {c.name: c for c in CELLS}
    missing = [n for n in names if n not in known]
    if missing:
        raise SystemExit(f"boot-matrix: no such cell {', '.join(missing)}")
    return [c for c in CELLS if c.name in names]
